### ISSUE TITLE
validate.exe linking fails with gcc/7.1.0

### ISSUE THREAD

**Issue by noraabiakar** [arbor-sim/arbor#465]
Linking error when building validate.exe using gcc/7.1.0
```
$ module load CMake/3.6.2
$ module load gcc/7.1.0
$ module swap PrgEnv-cray/6.0.4 PrgEnv-gnu
$ export CC=`which gcc`
$ export CXX=`which g++`
$ cmake ..
$ make -j 4

`_ZNSt17_Function_handlerIFbN3arb16cell_member_typeEEZNS0_L9one_probeES1_EUlS1_E_E9_M_invokeERKSt9_Any_dataOS1_' referenced in section `.rodata.cst8' of CMakeFiles/validate.exe.dir/validate_kinetic.cpp.o: defined in discarded section `.text._ZNSt17_Function_handlerIFbN3arb16cell_member_typeEEZNS0_L9one_probeES1_EUlS1_E_E9_M_invokeERKSt9_Any_dataOS1_[_ZN3arb23convergence_test_runnerIfE3runERNS_10simulationEffffRKSt6vectorIfSaIfEE]' of CMakeFiles/validate.exe.dir/validate_kinetic.cpp.o
`_ZNSt17_Function_handlerIFbN3arb16cell_member_typeEEZNS0_L9one_probeES1_EUlS1_E_E9_M_invokeERKSt9_Any_dataOS1_' referenced in section `.rodata.cst8' of CMakeFiles/validate.exe.dir/validate_synapses.cpp.o: defined in discarded section `.text._ZNSt17_Function_handlerIFbN3arb16cell_member_typeEEZNS0_L9one_probeES1_EUlS1_E_E9_M_invokeERKSt9_Any_dataOS1_[_ZN3arb23convergence_test_runnerIiE3runERNS_10simulationEifffRKSt6vectorIfSaIfEE]' of CMakeFiles/validate.exe.dir/validate_synapses.cpp.o
/usr/bin/ld: link errors found, deleting executable `../validate.exe'
```

**Comment by bcumming** [comment]
Which use of `static` are you referring too? The `-static` flag you say was added by the compiler wrapper, which works.

**Comment by halfflat** [comment]
The `static` keyword attached to the inline definition of `one_probe`.

**Comment by bcumming** [comment]
This bug isn't reported very often online. Some people fix it by changing the order of object files on the linker, others fix it by disable IPO (which makes sense, if the linker erroneously removes a symbol).

**Comment by bcumming** [comment]
https://stackoverflow.com/questions/9811503/unusual-c-linker-error-defined-in-discarded-section
https://mcuoneclipse.com/2017/11/12/solving-problem-with-gnu-linker-and-referenced-in-section-defined-in-discarded-section-error-message/

**Comment by halfflat** [comment]
Okie. Despite learning quite a bit more than I ever thought necessary about gcc, LTO, ELF function sections and what-not, I still haven't been able to rig up a MWE. Nonetheless, I'm pretty (somewhat?) sure it's a bug in gcc or the gcc toolchain as found on Piz Daint.

The fact that it's just that the wrapper includes `-march=core-avx2` that makes the difference between success and failure indicates it is quite brittle.

I'd recommend that the `static` keyword be removed from `one_probe`, and cross our fingers.

**Comment by bcumming** [comment]
The other option is to decorate the `one_probe` with `__attribute__((used))`, or its proper C++11 equivalent.

**Comment by halfflat** [comment]
The C++11 equivalent would be `[[gnu::used]]` prefixing the `static inline cell_member_predicate one_probe(cell_member_type pid)` definition. And adding this does work! But note that this wasn't the function that needed to be kept, but rather the `_M_invoke` method defined ultimately within the type-erasure implementation class within `std::function`, templated on the lambda type.

Essentially, it causes the desired outcome, but by dodging the brittleness of the underlying bug rather than by specifying directly that the troublesome function should be kept.

**Comment by noraabiakar** [comment]
Fix implemented in #747
